### ISSUE TITLE
Proposal: Lighthouse accessibility report should check for inaccessible scrolling pages and regions

### ISSUE THREAD

**Issue by daiplusplus** [GoogleChrome/lighthouse#11292]
#### Provide a basic description of the audit

This audit would inspect a page for any scrollable elements where browser/system-provided scrolling mechanisms have been disabled, replaced or restyled in a way that hinders the accessibility and ease-of-use for a site.

#### How would the audit appear in the report? 

I am proposing three related audits:

Two "Best practices" audits that inspect the page's CSS styling and rendered element content boxes:

1.  The first BP audit is rather strict, and would *recommend against making any changes to browser-provided scrolling mechanisms*.
    * This BP audit is raised (fails?) whenever it detects any element `el` where `el.scrollHeight > el.clientHeight` AND where `getComputedStyle( el ).overflow === 'none'` - or if any CSS rule is applied to any element that targets any of the `::-webkit-scrollbar` psedo-elements.

2. The second BP audit would permit the restyling of scrollbars provided their new styles meet necessary ease-of-use criteria:
    * Fitt's Law: The scrollbar track and thumb must be sufficiently large enough to easily target with a mouse cursor or touch input device (ideally based on a CSS media query).
    * If the scrollbar thumb is variable-height then the minimum size of the thumb must still be large enough to comfortably target.
        * Google's Angular documentation site is guilty of this - it's very difficult to click and drag the narrow scrollbar with a mouse: 
![image](https://user-images.githubusercontent.com/1693078/90758214-8523b080-e293-11ea-9626-a69d4d36e6b2.png)

    * Ideally the scrollbar track and thumb must have sufficient visual contrast (incl. contrast for colour-blind users) from each other, in addition to having sufficient contrast from the document body.
        * This could be difficult to automate: for example, there's an unfortunate bug in macOS Safari's scrollbars where if a page has a dark background in places (but is still overall light) then Safari will render a dark scrollbar thumb, which will become invisible or very difficult to see when the thumb overlaps the dark portion of the page. This problem exists because Safari no-longer renders scrollbar tracks, only thumbs. If Safari rendered a contrasting track then the thumb would always be visible.

A third audit would be a simple "Additional item to manually check" notice to remind users to ensure their website is accessible to users without scroll input devices (such as touchpads, mice scrollwheels, touchscreens) or users who otherwise can only use on-screen scrollbars to scroll content.

#### How would the test look when passing?

Green? 

####  Would there be additional details available?

If the audit fails due to a style-rule hiding scrolling UI or altering `::-webkit-scrollbar` then a jump-to-source link would be useful. As well as a list of elements that failed the audit.

#### How would the test look when failing?

Red? Orange?

####  What additional details are available? 

None

####  If the details are tabular, what are the columns?

N/A

####  If not obvious, how would passing/failing be defined? 

I briefly defined a test criteria above.

#### How is this audit different from existing ones?

I checked the audit-list and searched the GitHub Issues page and couldn't find any current audits that cover the accessibility of scrolling.

#### What % of developers/pages will this impact? 

This audit will affect 100% of pages that think they're cool by hiding the ugly Windows scrollbar because their webpages were designed by cool kids running macOS who think mouse-draggable scrollbars are for _losers_.

...while my above remark is firmly facetious, I note that because macOS remains the platform of choice for web-designers it is common for those web-designers to want to _export_ macOS's interaction concepts to the web (because they're genuinely cool, or novel, or work well... on Apple hardware), but they often are inaccessible or otherwise much more difficult to use without the requisite hardware, and this extends far beyond scrollbars.

Ultimately the root cause is web-designers not testing their designs fully on other platforms: while testing in Chrome and Safari on macOS might seem like good coverage for differences in browsers' default styles or for the responsive-mode or mobile-device emulation mode - but unfortunately Chrome/Chromium doesn't offer any kind of _desktop emulation_ mode (e.g. using Windows' default font sizes/metrics, always-visible wide scrollbars, etc) for users on other platforms.

#### How is the new audit making a better web for end users?

It would mean more web-pages will be accessible to users who cannot use - or do not possess - scrollwheels or touchpads.

#### What is the resourcing situation? 

I need to earn some karma - if it isn't too much work (say more than 20 hours?) I'm happy to implement this audit myself.

#### Any other links or documentation that we should check out?

Interestingly I couldn't find any direct references or mentions of scrollbars or scrolling-accessibility in WCAG 2.1 - which is odd, but there are plenty of articles that reference WCAG when making recommendations about scrollbar accessibility:

* https://www.w3.org/WAI/WCAG21/quickref
* https://adrianroselli.com/2019/01/baseline-rules-for-scrollbar-usability.html

I suppose the concerns about scrollbar-accessibility could come under the "Pointer gestures" section?

* https://www.w3.org/WAI/WCAG21/quickref/?currentsidebar=%23col_overview#pointer-gestures

### Are you willing to work on this yourself? 

Yes - provided there's a well-documented onboarding process.

### What is the motivation or use case for changing this?

My *personal* motivation is that my Windows desktop computer's mouse's scrollwheel suddenly stopped working recently (I assume some lint got stuck inside the wheel rotation sensor) which meant I could only scroll webpages by dragging a scrollbar. I came across a few webpages that had hidden their scrollbars entirely (I contacted the author of one website who said he did it because he felt it made his already minimalist site design look even cleaner - but when I explained to him it was making it impossible for me to read his blog's content he agreed to bring back scrollbars to his website).

**How is this beneficial to Lighthouse?**

It makes Lighthouse more comprehensive.

**Comment by patrickhulce** [comment]
Thanks for the detailed proposal @Jehoel! Have you brought this up with [axe-core](https://github.com/dequelabs/axe-core/) as a potential accessibility check? If it gets implemented there Lighthouse and many other a11y integrations would be able to pick it up easily :)

**Comment by Jehoel** [comment]
@patrickhulce I have not - I'll admit I didn't even know about axe-core until you mentioned it.

I see Chrome starting using axe-core in 2017 - I do see far more rules in axe-core than in Lighthouse's generated report - what is the extent of the integration and does Chrome/Chromium apply any filters to axe-core's rules?

**Comment by patrickhulce** [comment]
Yep!

https://github.com/GoogleChrome/lighthouse/blob/edecc71715bc3dbb99fbc1a4ed9dcbce424193dd/lighthouse-core/gather/gatherers/accessibility.js#L24-L54

**Comment by patrickhulce** [comment]
@Jehoel did you happen to bring this up with axe-core after our discussion with an issue we can link to? I'll mark this as waiting-on-upstream for now, but if they aren't aware of it then we'll be waiting for quite some time... :)

**Comment by Jehoel** [comment]
@patrickhulce Yes, I filed a proposal here: https://github.com/dequelabs/axe-core/issues/2538

**Comment by paulirish** [comment]
If there's renewed interest in this, pinging the axe-core issue is the right place to bring this back.
